cgp/individual.py

```diff
@@ -1,43 +1,84 @@
+import numpy as np
+from typing import Callable, List, Union
+
+try:
+    import sympy  # noqa: F401
+    from sympy.core import expr as sympy_expr  # noqa: F401
+
+    sympy_available = True
+except ModuleNotFoundError:
+    sympy_available = False
+
+try:
+    import torch  # noqa: F401
+
+    torch_available = True
+except ModuleNotFoundError:
+    torch_available = False
+
 from .genome import Genome
 from .cartesian_graph import CartesianGraph
 
 
-class Individual:
-    """An individual representing a particular computational graph.
+def mutate_genome(genome: Genome, mutation_rate: float, rng: np.random.RandomState) -> bool:
+    """Mutate a given genome.
+
+    Parameters
+    ----------
+    mutation_rate : float
+        Proportion of genes to be mutated, between 0 and 1.
+    rng : numpy.RandomState
+        Random number generator instance.
+
+    Returns
+    -------
+    bool
+        Whether all mutations were silent.
     """
+    n_mutations = int(mutation_rate * len(genome.dna))
+    assert n_mutations > 0
+
+    graph = CartesianGraph(genome)
+    active_regions = graph.determine_active_regions()
+    only_silent_mutations = genome.mutate(n_mutations, active_regions, rng)
+
+    return only_silent_mutations
 
-    def __init__(self, fitness, genome):
+
+class IndividualBase:
+    """Base class for all individuals.
+    """
+
+    def __init__(self, fitness: Union[float, None], genome: Union[Genome, List[Genome]]) -> None:
         """Init function.
 
         fitness : float
             Fitness of the individual.
-        genome: Genome instance
-            Genome of the individual.
         """
         self.fitness = fitness
-        self.genome = genome
-        self.idx = None  # an identifier to keep track of all unique genomes
+        self.genome: Union[Genome, List[Genome]]
+        self.idx: int
 
     def __repr__(self):
         return f"Individual(idx={self.idx}, fitness={self.fitness}, genome={self.genome}))"
 
-    def clone(self):
+    def clone(self) -> "IndividualBase":
         """Clone the individual.
 
         Returns
         -------
         Individual
         """
-        return Individual(self.fitness, self.genome.clone())
+        return type(self)(self.fitness, self.genome)
```

Comment:
```suggestion
        raise NotImplementedError()
```

Reply:
it does not make sense to use an instance of `IndividualBase`, does it? so i would just not provide any implementation for functions that are overridden in derived classes

Reply:
the idea is that we don't need to define the `clone` function in the child classes.